### DISCUSSION TITLE
Why are dataframe inserts so fast compared to Python list inserts?

### DISCUSSION THREAD

**Discussion by vedantroy** [duckdb/duckdb#6360] [Q&A]
Where `names` is a list of strings, inserting by first converting the list to a dataframe & then inserting the data into the table via the dataframe is quite fast (~ 0.5 seconds for ~ 300K items), whereas directly insert the items (even with `executemany`) takes so long, and just kill the script.

My questions are:
1. Why is inserting by converting to a dataframe first so much faster?
2. When I insert by converting to a dataframe, is the data actually being inserted into the table, or is duckdb just acting as a SQL-wrapper around the dataframe?

```
db = duckdb.connect(':memory:')
db.execute("CREATE TABLE frame (name VARCHAR UNIQUE)")
db.execute("CREATE INDEX frame_name_idx ON frame (name)")

# create a df with one column
df = pd.DataFrame(names, columns=['name'])

# use the df to insert
db.executemany("INSERT INTO frame VALUES (?)", [(name,) for name in names])
db.execute("INSERT INTO frame SELECT * from df")
```

**Comment by l1t1** [comment]
insert .. values (?) inserts one by one
insert .. select from inserts in batch

**Comment by l1t1** [comment]
insert into frame select 'a'||a from range(10)t(a);
Run Time (s): real 0.008 user 0.004075 sys 0.000000

**Comment by hannes** [comment]
so 1) its so much faster because instead of one Python operation per value, we have one for all of them and then DuckDB can read the entire column independently. 2) it actually inserts, but you can also directly query data frames, like `SELECT * FROM df` in your example

**Comment by FrancoisLepoutre** [comment]
Hi VedantRoy,

As a pythonista, I find that Duckdb is a wonderful tool. Exceptional in many respects. But it not always the best fit. One essential issue here is data persistent. How persistent are your data?

In the specific business cases that we are handling here, it fits a semi-temporary "in-memory" business object that we create from scratch from durable parquet resources any time a user initiate a "session". Takes from a fraction of second to a couple of them of course! But that's a cost we are ready to pay. And there is a reward: the data, even when big, stays safe in memory (and possibly out of it when memory is a tad short) as long as it it needed, ready for tapping.

We found that feeding duckdb via a combination of pyarrow and duckdb was really fast enough. Once the data is fed, it can SQL-mangled straight from the duck engine. And that's great. Of course when it comes to massive data-centric operations, there are situations where offloading to other data engines is more effective. That's in direct consequence of the "columnar" nature of our great duck! 

When the data operation can only be described as massive sets of array-centric operations, sql engines are possibly not always the best fit. But they are a wonderful base to tap from as you can download really any size of filtered subsets into numpy(+numba and its sets of accelerators), pandas and, now, rust-powered polars structures so fast, I find this fast upload/download from frame-based third parties functionality to be of exceptional value:-)

Of course, should your data justify fully persistent sets of duckdb native database file format, then the value would certainly be even greater. But first that depends on how persistent your big matrices may be. And of course, as you are possibly aware, the duckdb file format is not fully stabilized yet. And we need to wait a bit to use it as a long term storage. But the project is really going fast. It may not be THAT long :-)

**Comment by l1t1** [comment]
In my test, concat string to build a sql that include many values is faster than db.execute(insert .. values() 
```python
import duckdb
db = duckdb.connect(':memory:')
db.execute("SET enable_progress_bar=false")
db.execute("CREATE TABLE frame (name VARCHAR UNIQUE)")
db.execute("CREATE INDEX frame_name_idx ON frame (name)")

names=['a'+str(i) for i in range(100000)]
import time
t=time.time()
db.executemany("INSERT INTO frame VALUES (?)", [(name,) for name in names])
t1=time.time()
print(t1-t)
#37.85908555984497
db.execute("delete from frame")
t=time.time()
db.execute("insert into frame values('"+str.join("'),('",names)+"')")
t1=time.time()
print(t1-t)
#2.277604103088379
import pandas
db.execute("delete from frame")
t=time.time()
df=pandas.DataFrame(names, columns=['name'])
db.execute("INSERT INTO frame SELECT * from df")
t1=time.time()
print(t1-t)
#0.5772008895874023
```

**Comment by l1t1** [comment]
insert from csv is better
```python
db.execute("copy (select * from frame)to 'd:/frame.csv'")
db.execute("delete from frame")
t=time.time()
df=pandas.DataFrame(names, columns=['name'])
db.execute("INSERT INTO frame SELECT * from 'd:/frame.csv'")
t1=time.time()
print(t1-t)
#0.7176010608673096
```

**Comment by andreamoro** [comment]
## Still an issue in v1.4.2 — 8+ minutes vs 0.05s for 76K rows

I ran into this while building a personal NLP pipeline project. The pipeline persists term frequency data to DuckDB — 382 documents x ~200 terms each = ~76K rows. Using `executemany`, persistence took **over 8 minutes**. Switching to `read_csv()` brought it down to **under a second**.

Spent hours debugging, suspecting batch sizes, DELETE operations, table bloat, or Python loop overhead — none of which were the actual problem.

### Environment

- DuckDB: **1.4.2** (Python client)
- Python: **3.13.9**
- OS: **Ubuntu Linux 6.17.0-8-generic, x86_64**

### Reproducible benchmark

```python
import time, os, csv, tempfile, duckdb

# Generate 76,400 rows (382 docs x 200 terms)
rows = [
    (f'doc_{i}', f'term_{j}', j + 1, (j + 1) / 500.0)
    for i in range(382) for j in range(200)
]
cols = ['document_id', 'term', 'raw_count', 'tf_score']
print(f"Rows: {len(rows)}")

db = '/tmp/test_executemany.duckdb'
schema = 'CREATE TABLE test (document_id TEXT, term TEXT, raw_count INTEGER, tf_score DOUBLE)'

# --- Method 1: executemany ---
if os.path.exists(db): os.remove(db)
conn = duckdb.connect(db)
conn.execute(schema)
t0 = time.perf_counter()
conn.executemany('INSERT INTO test VALUES (?, ?, ?, ?)', rows)
print(f"executemany:      {time.perf_counter() - t0:.2f}s")
conn.close()

# --- Method 2: executemany in 382 batches of 200 ---
if os.path.exists(db): os.remove(db)
conn = duckdb.connect(db)
conn.execute(schema)
t0 = time.perf_counter()
for i in range(382):
    batch = rows[i * 200:(i + 1) * 200]
    conn.executemany('INSERT INTO test VALUES (?, ?, ?, ?)', batch)
print(f"executemany x382: {time.perf_counter() - t0:.2f}s")
conn.close()

# --- Method 3: CSV temp file + read_csv ---
if os.path.exists(db): os.remove(db)
conn = duckdb.connect(db)
conn.execute(schema)
t0 = time.perf_counter()
with tempfile.NamedTemporaryFile(mode='w', suffix='.csv', delete=False, newline='') as f:
    csv_path = f.name
    writer = csv.writer(f)
    writer.writerow(cols)
    writer.writerows(rows)
conn.execute(f"INSERT INTO test SELECT * FROM read_csv('{csv_path}', header=true)")
os.remove(csv_path)
print(f"CSV + read_csv:   {time.perf_counter() - t0:.2f}s")
conn.close()

os.remove(db)
```

### Results

| Method | Time | Factor |
|---|---|---|
| `executemany` (single call, 76K rows) | **16.08s** | 1x |
| `executemany` (382 calls x 200 rows) | **16.90s** | ~1x |
| CSV temp file + `read_csv()` | **0.05s** | **~320x faster** |

Batching doesn't help — `executemany` is equally slow whether called once with 76K rows or 382 times with 200 rows. The overhead is clearly per-row, not per-call.

### Real-world context

In my pipeline, the full persistence cycle (serialize + DELETE existing + `executemany` INSERT) for 382 documents took **8 minutes 37 seconds** wall clock. After switching to CSV-based insertion, the same operation completes in **~3 seconds** including serialization, DELETE, CSV write, and `read_csv` INSERT.

### Workaround

Write data to a temp CSV and use `read_csv()`. It uses DuckDB's native vectorized scanner instead of the row-by-row Python API path. No extra dependencies — just stdlib `csv` and `tempfile`.

**Reply by Alex-Monahan** [reply]
Well, there will always be a fairly substantial speed improvement when using a dataframe. Dataframes use C++ (Pandas) or Rust (Polars), so DuckDB can interface with them much more efficiently (without Python object overhead). 

A slightly simpler/faster workaround would be to use a StringIO object to hold the csv file in memory instead of writing it to a temp file.

**Reply by andreamoro** [reply]
I get that, but I wanted less dependency as possible, hence the reason of a CSV ... also to persist and avoid loosing anything in between in case of a crash.
Yet, I think this should be somehow reconsidered as a powerful db for doing analysis that cannot inject at a good speed on its own it doesn't make super sense to me.

**Comment by FrancoisLepoutre** [comment]
Hi all,

This thread tackles with an issue which I feel is more of a python question than a generic duckdb one. It looks more to a question of dependencies for python-based applications that rely on our great duck. 

Upload data from external file resources into a duckdb db is literally a snap now! The parquet feeder which is part and parcel of the the duckdb monolith is a joy to work with and deploying such duckdb-based python app is a treat!  However integrating other big data libraries in a python app may not be such primrose path.

There are certainly effective duckdb data uploading/downloading solutions within the python ecosystem that allow feeding in-memory resources into duckdb. Although I am no techie, I tested pyarrow and definitely liked it. However I understand that Pandas is still the main game in python data science city although you may also address the issue with Polars as well. But I caught it that deploying Pandas just to feed our duck may not justify the technical debt, especially when you use is for relatively benign operations. 

An open question now: "Can you use a reduced subset of Polars, reduced in terms of size essentially, as a more compact alternative to Pandas for effective data upload/download?"

**Reply by andreamoro** [reply]
Hi Francois
Nobody is saying duckdb is not great. It is my default columnar dB option since a while.

All I'm dating is that the driver, perhaps the one for python, might require some attention because or doesn't make sense to have something great that fails on repetitive inserts.

For now, I've relaxed the requirement and I'm passing via a csv import. But I'd love not to have even that extra operation in all honesty.

**Comment by FrancoisLepoutre** [comment]
Andrea,

_>something great that fails on repetitive inserts._

Thanks for the feedback and the exchange on this thread. I have run into the same kind of speed issues when testing the same kind of arbitrary imports via execute python instructions. My POV is: 
1. The duckdb engine architecture is OLAP.  It definitely favors large operations over small ones; duckdb is no sqlite in terms of data feeding,
2. In such a setup, large sets of repetitive calls to atomic read or write operations have a high processing cost,
3. Conversely, a single call that transfer large sets of data can be high-speed, especially on zero-copy operations; this is where our duck shines.

Pandas and arrow data (via pyarrow) "bulk feeding" operations can be impressively fast, especially when they take advantage of zero-copy. But the perf levels one can achieve with arbitrary standard executemany instructions can only be irremediably low. Possibly improvements could reached. But, IMVHO, not anything that would make them an decent alternative for anything largish. 

I may be wrong of course and I'd be glad to be proven wrong. By the way running your upload through csv-s (or json) strings looks like a decent intermediate solution! Especially, as Alex mentioned. if you run it via StringIO entries.
